source/common/upstream/upstream_impl.cc

```diff
@@ -590,11 +590,11 @@ void HostSetImpl::updateHosts(PrioritySet::UpdateHostsParams&& update_hosts_para
   rebuildLocalityScheduler(healthy_locality_scheduler_, healthy_locality_entries_,
                            *healthy_hosts_per_locality_, healthy_hosts_->get(), hosts_per_locality_,
                            excluded_hosts_per_locality_, locality_weights_,
-                           overprovisioning_factor_);
+                           overprovisioning_factor_, seed);
```

Comment:
One other question that came up while reading this PR. We're threading seed around in a lot of places. Why can't it live as a constant (for the LB) for the cluster or even server context?

Reply:
That's a good question.
While working on this PR I had the following questions on my mind:
1. What is the granularity of the seed (e.g., is it for a single cluster, or for all clusters)?
2. Should the seed be kept in memory (say a data-member of some class)? for example initialized once and then used throughout the lifetime of the process.
My observation was that using a constant seed (let's say throughout the duration of an Envoy process) will end up with a wrong randomization. Specifically for low-QPS scenarios where EDS updates do not change the weights drastically, one may end up with an incorrect traffic pattern.
I've also decided against keeping the seed in memory (data member), but rather pass it as an argument to the function, as I'm not sure if "spending" memory for the seed is necessary. Moreover, adding it to the `HostSetImpl` class and passing using it's c'tor would have probably ended up with a similar "plumbing" work.

I think the current approach allows the code to be changed later and the seed can be saved in some context and passed just as it is being passed now.

If the concern is the number of parameters for the function (which is indeed large), then I wonder whether the function needs to be refactored somehow by splitting it to multiple ones, or if the arguments should be passed in a struct (IMHO the former is the better approach, and the latter doesn't really solve the problem).

Reply:
It's probably fine, just lots of plumbing. Independent of this, it might make sense to move some of these extensively plumbed params that are appearing in many places to an options struct.